File: 2021/cpp/Day_2.cpp
#include <fstream>
#include <iostream>
#include <vector>

void part_1();
void part_2();
std::vector<std::string> parse_input();

int main() {
  part_1();
  part_2();
  return 0;
}

void part_1() {
  std::vector<std::string> instructs = parse_input();
  int horizontal = 0;
  int vertical = 0;

  for (int i = 0; i < instructs.size(); i++) {
    std::string line_i = instructs[i];
    int spacer = line_i.find(" ");

    std::string dir = line_i.substr(0, spacer);
    int num = stoi(line_i.substr(spacer + 1, (line_i.length() - spacer - 1)));

    if (dir == "up") {
      vertical -= num;
    } else if (dir == "down") {
      vertical += num;
    } else if (dir == "forward") {
      horizontal += num;
    }
  }

  int final_s = horizontal * vertical;
  std::cout << "The final answer for Part 1 is: " << final_s << std::endl;
}

void part_2() {
  std::vector<std::string> instructs = parse_input();
  int horizontal = 0;
  int vertical = 0;
  int aim = 0;

  for (int i = 0; i < instructs.size(); i++) {
    std::string line_i = instructs[i];
    int spacer = line_i.find(" ");

    std::string dir = line_i.substr(0, spacer);
    int num = stoi(line_i.substr(spacer + 1, (line_i.length() - spacer - 1)));

    if (dir == "up") {
      aim -= num;
    } else if (dir == "down") {
      aim += num;
    } else if (dir == "forward") {
      horizontal += num;
      vertical += (aim * num);
    }
  }

  int final_s = horizontal * vertical;
  std::cout << "The final answer for Part 1 is: " << final_s << std::endl;
}

std::vector<std::string> parse_input() {
  std::vector<std::string> nums;
  std::ifstream f("Inputs/day_2_input.txt");

  std::string s;

  while (getline(f, s)) {
    nums.push_back(s);
  }

  f.close();

  return nums;
}

File: 2021/Python/Day_2.py
def parse_input():
    # Parse data from txt file
    total_arr = []

    with open("Inputs/day_2_input.txt") as input_file:
        for line in input_file:
            total_arr.append(line)

    return total_arr


def part_1():
    ins = parse_input()
    horizontal = 0
    vertical = 0
    for i in ins:
        x = i.split(" ")
        if x[0] == "up":
            vertical -= int(x[1])
        elif x[0] == "down":
            vertical += int(x[1])
        elif x[0] == "forward":
            horizontal += int(x[1])

    print("The final result for Part 1 is: " + str(horizontal * vertical))


def part_2():
    ins = parse_input()
    horizontal = 0
    vertical = 0
    aim = 0

    for i in ins:
        x = i.split(" ")
        if x[0] == "up":
            aim -= int(x[1])
        elif x[0] == "down":
            aim += int(x[1])
        elif x[0] == "forward":
            horizontal += int(x[1])
            vertical += aim * int(x[1])

    print("The final result for Part 2 is: " + str(horizontal * vertical))


if __name__ == "__main__":
    part_1()
    part_2()
